lib/fined.js

```diff
@@ -0,0 +1,151 @@
+const fs = require('fs');
+const path = require('path');
+
+const isString = require('lodash.isstring');
+const isArray = require('lodash.isarray');
+const isPlainObject = require('lodash.isplainobject');
+const isEmpty = require('lodash.isempty');
+const pick = require('lodash.pick');
+
+const expandTilde = require('expand-tilde');
+const parsePath = require('parse-filepath');
+
+var Fined = function(basedir, defaultExtensions) {
+  Constructor.prototype = Fined.prototype;
+  return new Constructor(basedir, defaultExtensions);
+
+  function Constructor(basedir, exts) {
+    this.basedir = path.resolve(isString(basedir) ? basedir : '.');
+    this.defaultExtensionArray = createExtensionArray(exts);
+    this.defaultExtensionMap = createExtensionMap(exts);
+  }
+};
+
+Fined.prototype.find = function(searchPath, name) {
+  var expandedArr = expandPath.call(this, searchPath, name);
+  for (var i = 0, n = expandedArr.length; i < n; i++) {
+    var found = findWithExpandedPath(expandedArr[i]);
+    if (found) {
+      return found;
+    }
+  }
+  return null;
+};
+
+function expandPath(searchPath, name) {
+  var extArr = this.defaultExtensionArray, extMap = this.defaultExtensionMap;
+  var filePath, findUp = false;
+
+  if (isString(searchPath)) {
+    filePath = searchPath;
+
+  } else if (isPlainObject(searchPath) && isString(searchPath.path)) {
+    filePath = searchPath.path;
+    findUp = !!searchPath.findUp;
+
+    if ('extensions' in searchPath) {
+      extArr = createExtensionArray(searchPath.extensions);
+      extMap = createExtensionMap(searchPath.extensions);
+    }
+  } else {
+    return [];
+  }
+
+  filePath = expandTilde(filePath);
+
+  if (isString(name) && name.length > 0) {
+    filePath = path.join(filePath, name);
+  }
+
+  return extArr.map(createExpandedPath.bind(this));
+
+
+  function createExpandedPath(ext) {
+    var expanded = {
+      path: filePath + ext,
+      basedir: this.basedir,
+      findUp: findUp,
+      extension: (extMap ? pick(extMap, ext) : ext),
+    };
+
+    var parsed = parsePath(expanded.path);
+    if (parsed.isAbsolute) {
+      expanded.findUp = false;
+      expanded.path = expanded.path.slice(parsed.root.length);
+      expanded.basedir = parsed.root;
+
+    } else if (parsed.root) { // expanded path has a drive letter on Windows.
+      expanded.path = expanded.path.slice(parsed.root.length);
+      expanded.basedir = path.resolve(parsed.root);
+    }
+
+    return expanded;
+  }
+}
+
+function findWithExpandedPath(expanded) {
+  var found = expanded.findUp ?
+    findUpFile(expanded.path, expanded.basedir) :
+    findFile(path.resolve(expanded.basedir, expanded.path));
+
+  if (!found) {
+    return null;
+  }
+  return { path: found, extension: expanded.extension };
+
+
+  function findFile(filepath) {
```

Comment:
I would lift this out of this function--it doesn't rely on any state and doesn't need to be created dynamically with every invocation of `findWithExpandedPath`